working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+
+It is a compile-time error for an extension declaration to start with the
+keyword `protected`, unless it is explicit (*that is, unless it also has
+the keyword `type`*).
+
+*A protected extension type should not allow for implicit extension method
+invocations, because they are inherently not guarded by the execution of an
+extension type constructor.*
+
+*For example:*
+
+```dart
+protected extension type nat on int {
+  factory nat(int value) =>
+      value >= 0 ? value : throw "Attempt to create an invalid nat";
+}
+
+void main() {
+  nat n1 = 42; // Error.
+  var n2 = nat(42); // OK at compile time, and at run time.
+  var n3 = nat(-1); // OK at compile time, throws at run time.
+}
+```
+
+*The following example illustrates the subtyping relationships; it
+illustrates the use of a constructor to enforce an invariant (that the
+`int` in the `IntBox` is even); and it illustrates that the available
+methods (there is just one: `next()`) all preserve that invariant.*
+
+```dart
+class IntBox {
+  int i;
+  IntBox(this.i);
+}
+
+protected extension type EvenIntBox on IntBox {
+  factory EvenIntBox(int i) =>
+      i.isEven ? IntBox(i) : throw "Invalid EvenIntBox";
+  factory EvenIntBox.fromIntBox(IntBox intBox) =>
+      intBox.i.isEven ? intBox : throw "Invalid EvenIntBox";
+  void next() => this.i += 2;
+}
+
+void main() {
+  var evenIntBox = EvenIntBox(42);
+  evenIntBox.next(); // Methods of `EvenIntBox` maintain the invariant.
+  evenIntBox = intBox; // Compile-time error, types not assignable.
+  evenIntBox = intBox as EvenIntBox; // Compile-time error, can't cast.
+
+  // We cannot escape by a cast when the protected extension type is
+  // a type argument (or a return/parameter type in a function type).
+  var evenIntBoxes = [evenIntBox]; // Type `List<EvenIntBox>`.
+  evenIntBoxes[0].next(); // Elements typed as `EvenIntBox`.
+  List<IntBox> intBoxes = evenIntBoxes; // Compile-time error.
+  intBoxes = evenIntBoxes as dynamic; // Run-time error.
+
+  // We _can_ escape the protected extension type by an explicit cast.
+  var intBox = evenIntBox as IntBox; // OK statically and dynamically.
+  intBox.i++; // Invariant of `evenIntBox` violated!
+}
+```
+
+*Note that an explicit cast can be used to escape the protected extension
+type and obtain a reference to the underlying object under some other type,
+e.g., the on-type. This means that we can break the invariant, because the
+object is no longer handled with the discipline that the extension type
+members apply.*
+
+*Hence, the protection offered by a protected extension type is easy to
+violate, but the point is that it is reasonably easy to avoid violating
+this protection, and hence the mechanism can be used by developers who wish
+to maintain that specific discipline.*
+
+*A harder protection can be achieved by boxing the extension type, as
+described in the next section.*
+
+
+### Boxing
+
+This section describes the implicitly induced `box` getter of an explicit
```

Comment:
I think the notion of boxing needs more discussion separately.  I suspect that if we want boxing, we may want a different approach.

Reply:
I created https://github.com/dart-lang/language/issues/1491 to handle this discussion separately.